include/envoy/http/filter.h

```diff
@@ -36,6 +36,25 @@ enum class FilterHeadersStatus {
   // results in creating a header only request/response.
   // This status MUST NOT be returned by decodeHeaders() when end_stream is set to true.
   ContinueAndEndStream,
+  // Continue iteration to remaining filters, but do not end the stream.
+  //
+  // Used when a filter wants to add a body to headers-only request/response, but this body is not
+  // readily available.
+  // This causes the headers iteration to continue, but the stream does not end.
+  // The filter is responsible to continue the stream by providing a body through
+  // injectEncodedDataToFilterChain()/injectDecodedDataToFilterChain().
+  // If the filter cannot provide a body the stream should be reset.
+  //
+  // Adding a body through calling addDecodedData()/addEncodedData() then
+  // continueDecoding()/continueEncoding() is currently not supported and causes an assert failure.
+  //
+  // TODO(yosrym93): Support adding a body in this case by calling addDecodedData()/addEncodedData()
+  // then continueDecoding()/continueEncoding(). To support this a new FilterManager::IterationState
+  // needs to be added and set when a filter returns this status in FilterManager::encodeHeaders.
+  // Currently, when a filter returns this, the IterationState is Continue. This causes ASSERTs at
+  // FilterManager::commonContinue() to fail when continueDecoding()/continueEncoding() is called;
+  // due to trying to continue iteration when the IterationState is already Continue.
+  ContinueAndDontEndStream,
```

Comment:
Generally with the docs here I'd say we should make sure that its clear how this differs from Continue (when would you use one over the other) and exactly when it can be used.

Reply:
In the docs above it says: 
```
Continue iteration to remaining filters, but do not end the stream.
Used when a filter wants to add a body to headers-only request/response, but this body is not readily available.
```
Do you this is not clear enough? Do you have suggestions in improving the wording here?

Reply:
I guess the my reading of it made it sound more of a suggested use of this return value than an ASSERT enforced invariant. Compare the wording with `ContinueAndEndStream` or `StopIteration` the language is a lot stronger, so it might be good to be consistent?

How about something along the lines of

```
Continue iteration to remaining filters, but delay ending the stream. This status MUST NOT be returned when end_stream is already set to false.

This value is intended for filters that want to add a body to a request/response that only contains headers, but this body is not readily available. Delaying end_stream allows the filter to add the body once it's available without blocking.
```

(Feel free to reword, just a suggestion)